### PR TITLE
Victor/eco 948/admin configuration rate limit

### PR DESCRIPTION
**Pull request by Doc999tor**
- [x ] linked to a JIRA issue
- [ ] description of the change
- [ ] tests for the newly added code
- [ ] if this is a breaking change, contains backward support code
- [ ] if this requires client changes or partner updates, updated dependendants
- [ ] were metrics and alerts added
- [ ] should I change documentation

### CODE REVIEW

File: scripts/src/admin/routes.ts [hunk 1]
```diff
@@ -25,286 +32,11 @@ function jsonResponse(func: (body: any, params: any, query: any) => Promise<stri
...
-		.flex-row-container {
-		    display: flex;
-		    flex-direction: row;
-		    flex-wrap: nowrap;
-		    justify-content: center;
-		    align-content: stretch;
-		    align-items: center;
-	    }
-
+		const html = (await readFile(path.join(__dirname, "../../src/admin/wrapper.html"), { encoding: "utf8" }))
```
Comment (nitzantomer) [inline scripts/src/admin/routes.ts]: maybe extract the file path to be a constant in the top of the file?

File: scripts/src/admin/services.ts [hunk 2]
```diff
@@ -689,3 +692,15 @@ export async function changeOffer(body: ChangeOfferData, params: { offer_id: str
 
 	await offer.save();
 }
+
+export const updateAppConfig = async function(req: Request, res: Response) {
+	const app = (await Application.findOne({ id: req.params.application_id }))!;
+	try {
+		app.config = req.body;
```
Comment (nitzantomer) [inline scripts/src/admin/services.ts]: maybe define a type for this specific request so it will be clear?

File: scripts/src/admin/wrapper.html [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+		<title>Marketplace Admin</title>
+		<style>
+		body {
+			font-family: Arial;
+			background: #ffffff;
+		}
+		td {
+			vertical-align: bottom;
+		}
+		tr:nth-child(even) {background: #FFFFFF}
```
Comment (nitzantomer) [inline scripts/src/admin/wrapper.html]: keep the conventions and break lines

File: scripts/src/admin/wrapper.html [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+		</style>
+		<style>
+		.flex-row-container {
+		    display: flex;
+		    flex-direction: row;
+		    flex-wrap: nowrap;
+		    justify-content: center;
+		    align-content: stretch;
+		    align-items: center;
+	    }
```
Comment (nitzantomer) [inline scripts/src/admin/wrapper.html]: indentation

File: scripts/src/admin/wrapper.html [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
```
Comment (nitzantomer) [inline scripts/src/admin/wrapper.html]: extra empty line

File: scripts/src/admin/wrapper.html [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+		.btn {
+			width: 52%;
+			height: 8.5%;
+			margin: 5px;
+			cursor: pointer;
+			border-radius: 9px;
+		}
+
+		.overlay {
+            position: fixed;
```
Comment (nitzantomer) [inline scripts/src/admin/wrapper.html]: indentation

File: scripts/src/admin/wrapper.html [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+			height: 80%;
+			top: 50%;
+			left: 50%;
+			padding: 2%;
+			color: black;
+			background-color: white;
+			transform: translate(-50%,-50%);
+			-ms-transform: translate(-50%,-50%);
+		}
+
```
Comment (nitzantomer) [inline scripts/src/admin/wrapper.html]: either we have line breaks between rules or not

File: scripts/src/admin/wrapper.html [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+		}
+		input {
+			width: 70px;
+		}
+		#footer {
+			height: 70px;
+			width: 100%;
+		}
+		</style>
+		<style>
```
Comment (nitzantomer) [inline scripts/src/admin/wrapper.html]: why not in the same style tag?

File: scripts/src/admin/wrapper.html [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+				overlayTextElement = document.querySelector(".overlay .text");
+				previewBtnElemnt = document.querySelector(".overlay .preview-btn");
+				previewElemnt = document.querySelector(".overlay .preview iframe");
+				publishBtnElemnt = document.querySelector(".overlay .publish-btn");
+
+				toastElemnt = document.querySelector("#toast");
+				previewBtnElemnt.addEventListener("click", refreshPreviewBtnHandler)
+				publishBtnElemnt.addEventListener("click", publishBtnHandler)
+			});
+
```
Comment (nitzantomer) [inline scripts/src/admin/wrapper.html]: too many empty lines

File: scripts/src/admin/wrapper.html [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+				overlayJsonEditor = new JSONEditor(overlayTextElement, jsonEditorOptions, JSON.parse(unescape(text), null,  2));
+				overlayJsonEditor.expandAll();
+			}
+			function overlayOff() {
+				overlayElement.style.display = "none";
+				overlayTextElement.textContent = "";
+				previewElemnt.src = "";
+			}
+
```
Comment (nitzantomer) [inline scripts/src/admin/wrapper.html]: either we have empty lines between functions or not

**Review by nitzantomer — CHANGES_REQUESTED** [review]

File: scripts/src/admin/routes.ts (hunk 1 above)
Comment (Doc999tor) [reply scripts/src/admin/routes.ts]: Can you pls guide me what Request type fits here? Or maybe I need to create one?

File: scripts/src/admin/services.ts (hunk 2 above)
Comment (Doc999tor) [reply scripts/src/admin/services.ts]: Can you pls guide me what Request type fits here? Or maybe I need to create one?

File: scripts/src/admin/routes.ts (hunk 1 above)
Comment (Doc999tor) [reply scripts/src/admin/routes.ts]: Moved out of the closure function

File: scripts/src/admin/wrapper.html (hunk 3 above)
Comment (Doc999tor) [reply scripts/src/admin/wrapper.html]: Added spaces and a semicolon

File: scripts/src/admin/services.ts (hunk 2 above)
Comment (nitzantomer) [reply scripts/src/admin/services.ts]: you can do:
```
export type UpdateAppConfigRequest = Request & {
	body: {
		...
	}
}

export const updateAppConfig = async function(req: UpdateAppConfigRequest, res: Response) {
	...
```

**Review by nitzantomer — APPROVED** [review]

File: scripts/src/admin/wrapper.html [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+			function attachEditorKeyboardEvents(editor, e) {
+				switch (e.keyCode) {
+					case 13: if (e.ctrlKey) { updateConfig(editor) } break;
+					case 27: closeEditor(editor); break;
+				}
+			}
+	</script>
+	</body>
+
+</html>
```
Comment (liorama) [inline scripts/src/admin/wrapper.html]: newline

File: scripts/src/admin/wrapper.html [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+				try {
+					jsonData = JSON.stringify(json);
+					return jsonData;
+				} catch {
+					toast("Invaid JSON");
+					return false;
+				}
+			}
+
+			function editAppConfigs(json, appId) {
```
Comment (liorama) [inline scripts/src/admin/wrapper.html]: Whats this? Is this needed?

Comment (liorama) [reply scripts/src/admin/wrapper.html]: @Doc999tor do we need both this function and the below `openEditor(jsonArea, appId)` function?

File: scripts/src/admin/wrapper.html [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+			}
+			function refreshPreview(json){
+				let jsonData = stringifyJson(json);
+				if (!jsonData){
+					return;
+				}
+				previewElemnt.src = "${ webview }?cacheBuster=${ now }&jsonData=" + encodeURIComponent(jsonData);
+			}
+
+			function openEditor(jsonArea, appId) {
```
Comment (liorama) [inline scripts/src/admin/wrapper.html]: I would split this into a function that handles the UI and one that just opens the editor (and accepts just text and element) and is called by the first function.

**Pull request merged**